llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13598,14 +13598,26 @@ struct NodeExtensionHelper {
     case RISCVISD::VZEXT_VL:
     case RISCVISD::FP_EXTEND_VL:
       return OrigOperand.getOperand(0);
+    case ISD::SPLAT_VECTOR: {
+      SDValue Op = OrigOperand.getOperand(0);
+      unsigned Opc = Op.getOpcode();
+      if (SupportsSExt && Opc == ISD::SIGN_EXTEND_INREG)
+        return Op.getOperand(0);
+
+      if (SupportsZExt && Opc == ISD::AND)
+        return Op.getOperand(0);
```

Comment:
Why do we need to look through and/sign_extend_inreg? Do they not automatically get folded away if we can SelectionDAG already knows `DAG.MaskedValueIsZero(Op, APInt::getBitsSetFrom(ScalarBits, NarrowSize))` or `DAG.ComputeMaxSignificantBits(Op) <= NarrowSize`?

Reply:
These 2 patterns correspond to the case (splat_vector (sext op)). But we may also need to consider a constant splat like (splat_vector -1).

Reply:
How come we don't need to do this for VMV_V_X_VL though?

Reply:
I presume VMV_V_X_VL is only introduced after type legalization so it's always the scalar op is always XLenVT. But for splat_vector we need to deal with non-xvlen scalar ops?

Reply:
I just noticed a mistake in the test: 
```
define <vscale x 8 x i64> @vwadd_vx_splat_zext(<vscale x 8 x i32> %va, i32 %b) {
; RV64-LABEL: vwadd_vx_splat_zext:
; RV64:       # %bb.0:
; RV64-NEXT:    vsetvli a1, zero, e32, m4, ta, ma
; RV64-NEXT:    vwaddu.vx v16, v8, a0
; RV64-NEXT:    vmv8r.v v8, v16
; RV64-NEXT:    ret
  %sb = zext i32 %b to i64
  %head = insertelement <vscale x 8 x i64> poison, i64 %sb, i32 0
  %splat = shufflevector <vscale x 8 x i64> %head, <vscale x 8 x i64> poison, <vscale x 8 x i32> zeroinitializer
  %vc = zext <vscale x 8 x i32> %va to <vscale x 8 x i64>
  %ve = add <vscale x 8 x i64> %vc, %splat
  ret <vscale x 8 x i64> %ve
}
```

Actually at RV64, the below is valid
```
define <vscale x 8 x i64> @vwadd_vx_splat_zext(<vscale x 8 x i32> %va, i32 %b) {
; RV64-LABEL: vwadd_vx_splat_zext:
; RV64:       # %bb.0:
; RV64-NEXT:    andi	a0, a0, -1
; RV64-NEXT:    vsetvli a1, zero, e32, m4, ta, ma
; RV64-NEXT:    vwaddu.vx v16, v8, a0
; RV64-NEXT:    vmv8r.v v8, v16
; RV64-NEXT:    ret
  %sb = zext i32 %b to i64
  %head = insertelement <vscale x 8 x i64> poison, i64 %sb, i32 0
  %splat = shufflevector <vscale x 8 x i64> %head, <vscale x 8 x i64> poison, <vscale x 8 x i32> zeroinitializer
  %vc = zext <vscale x 8 x i32> %va to <vscale x 8 x i64>
  %ve = add <vscale x 8 x i64> %vc, %splat
  ret <vscale x 8 x i64> %ve
}
```
It is due to this unnecessary pattern match. I will fix it soon.

Reply:
I checked this out locally, I see now that for some reason we aren't able to combine away this redundant and:

```
Optimized legalized selection DAG: %bb.0 'vwadd_vx_splat_zext:'
SelectionDAG has 15 nodes:
  t0: ch,glue = EntryToken
      t2: nxv8i32,ch = CopyFromReg t0, Register:nxv8i32 %0
          t4: i64,ch = CopyFromReg t0, Register:i64 %1
        t18: i64 = and t4, Constant:i64<4294967295>
      t22: nxv8i32 = splat_vector t18
      t20: nxv8i1 = RISCVISD::VMSET_VL Register:i64 $x0
    t23: nxv8i64 = RISCVISD::VWADDU_VL t2, t22, undef:nxv8i64, t20, Register:i64 $x0
  t15: ch,glue = CopyToReg t0, Register:nxv8i64 $v8m8, t23
  t16: ch = RISCVISD::RET_GLUE t15, Register:nxv8i64 $v8m8, t15:1
```

I think if we want to peek through it in getSource we should probably check which operand is the source and which is the constant. But it's probably better to teach DAGCombiner to fold this, or handle this ourselves as a target specific combine for splat_vector. Since I'm think simplifyDemandedBits should be able to catch this